source/Octopus.Tentacle/Communications/AutofacServiceFactory.cs

```diff
@@ -17,48 +17,48 @@ public class AutofacServiceFactory : IServiceFactory, IServiceRegistration, IDis
         // Must never be modified as it is required for backwards compatability in BackwardsCompatibleCapabilitiesV2Decorator
         const string TentacleServiceShuttingDownMessage = "The Tentacle service is shutting down and cannot process this request.";
         readonly ILifetimeScope scope;
-        readonly Dictionary<string, Type> serviceTypes = new();
+        readonly Dictionary<string, KnownService> knownServices = new();
 
         public AutofacServiceFactory(ILifetimeScope scope, IEnumerable<IAutofacServiceSource> sources)
         {
             this.scope = scope.BeginLifetimeScope(b =>
             {
-                foreach (var service in sources.SelectMany(x => x.ServiceTypes.EmptyIfNull()))
+                foreach (var knownService in sources.SelectMany(x => x.KnownServices.EmptyIfNull()))
                 {
-                    BuildService(b, service);
+                    BuildService(b, knownService);
                 }
             });
         }
 
-        void BuildService(ContainerBuilder builder, Type serviceType)
+        void BuildService(ContainerBuilder builder, KnownService knownService)
         {
-            var registrationBuilder = builder
-                .RegisterType(serviceType)
+            //register the service implementation AsSelf(), so we can resolve it in the future
+            builder
+                .RegisterType(knownService.ServiceImplementationType)
                 .AsSelf()
                 .SingleInstance();
 
-            var interfaces = serviceType.GetInterfaces();
-            if (serviceType.IsInterface || interfaces.IsNullOrEmpty())
-            {
-                throw new InvalidServiceTypeException(serviceType);
-            }
-            
-            foreach (var face in interfaces)
+            //track the contract type to their known service implementations
+            //the contract type is the one that is sent across the wire (typically an IService sync contract)
+            knownServices[knownService.ServiceContractType.Name] = knownService;
```

Comment:
Instead of having a dictionary of known services, can we just leverage the container and use [named services](https://autofac.readthedocs.io/en/latest/advanced/keyed-services.html#named-services) instead?
I do realise that you're following the pattern used prior to your change though.

Reply:
Oh wait, looks like it's needed for the `RegisteredServiceTypes` property.  We'd need a common service interface to retrieve them without that dictionary.  Bugger.
Never mind then, as you were!